flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/OverTypeStrategy.java

```diff
@@ -0,0 +1,197 @@
+
+        if (precedingErr.isPresent()) {
+            return callContext.fail(throwOnFailure, precedingErr.get());
+        }
+
+        final LogicalTypeRoot followingTypeRoot = following.getLogicalType().getTypeRoot();
+        final Optional<String> followingErr =
+                validateFollowingPreceding(
+                        followingTypeRoot, FollowingPreceding.FOLLOWING, callContext);
+
+        if (followingErr.isPresent()) {
+            return callContext.fail(throwOnFailure, followingErr.get());
+        }
+
+        final boolean isPrecedingRowWindow =
+                isRowWindow(callContext, precedingTypeRoot, FollowingPreceding.PRECEDING);
+        final boolean isFollowingRowWindow =
+                isRowWindow(callContext, followingTypeRoot, FollowingPreceding.FOLLOWING);
+
+        if (isPrecedingRowWindow && !isFollowingRowWindow
+                || !isPrecedingRowWindow && isFollowingRowWindow) {
+            return callContext.fail(
+                    throwOnFailure, "Preceding and following must be of same interval type.");
+        }
+
+        return Optional.of(argumentDataTypes);
+    }
+
+    private static boolean isRowWindow(
+            CallContext callContext,
+            LogicalTypeRoot precedingTypeRoot,
+            FollowingPreceding followingPreceding) {
+        switch (precedingTypeRoot) {
+            case BIGINT:
+                return true;
+            case SYMBOL:
+                final OverWindowRange windowRange =
+                        callContext
+                                .getArgumentValue(followingPreceding.pos, OverWindowRange.class)
+                                .get();
+                return windowRange == OverWindowRange.UNBOUNDED_ROW
+                        || windowRange == OverWindowRange.CURRENT_ROW;
+            default:
+                return false;
+        }
+    }
+
+    private enum FollowingPreceding {
+        PRECEDING("Preceding", 2),
+        FOLLOWING("Following", 3);
+
+        private final int pos;
+        private final String name;
+
+        FollowingPreceding(String name, int pos) {
+            this.name = name;
+            this.pos = pos;
+        }
+    }
+
+    private Optional<String> validateFollowingPreceding(
+            LogicalTypeRoot typeRoot,
+            FollowingPreceding followingPreceding,
+            CallContext callContext) {
+        switch (typeRoot) {
+            case BIGINT:
+                if (callContext.getArgumentValue(followingPreceding.pos, Long.class).get() <= 0) {
+                    return Optional.of(
+                            String.format(
+                                    "%s row interval must be larger than 0.",
+                                    followingPreceding.name));
+                }
+                break;
+            case INTERVAL_DAY_TIME:
+                if (callContext
+                                .getArgumentValue(followingPreceding.pos, BigDecimal.class)
```

Comment:
when is can be INTERVAL_DAY_TIME?

Reply:
When you say e.g. `2.minutes preceding CURRENT_RANGE`